ggshield/sca/output/text_handler.py

```diff
@@ -1,20 +1,150 @@
-from ggshield.core.text_utils import STYLE, file_info, format_text
+from ggshield.core.text_utils import STYLE, file_info, format_text, pluralize
 from ggshield.sca.collection import (
     SCAScanAllVulnerabilityCollection,
     SCAScanDiffVulnerabilityCollection,
 )
 from ggshield.sca.output.handler import SCAOutputHandler
 from ggshield.sca.sca_scan_models import (
     SCALocationVulnerability,
-    SCAVulnerability,
-    SCAVulnerablePackageVersion,
+    SCAScanAllOutput,
+    SCAScanDiffOutput,
 )
 
 
+class OutputIncidentData(NamedTuple):
+    package_full_name: str
+    version: str
+    severity: str
+    summary: str
+    fixed_version: Optional[str]
+    identifier: str
+    cve_ids: List[str]
+
+    def sort_key(self) -> int:
+        if self.severity == "critical":
+            return 0
+        if self.severity == "high":
+            return 10
+        if self.severity == "medium":
+            return 20
+        if self.severity == "low":
+            return 30
+        return 100
+
+
+class OutputLocationData(NamedTuple):
+    """
+    Aggregate SCA incidents data to be printed in one class
+    The counters allow to sort according to the number of incidents with each severity
+    """
+
+    location: str
+    added: List[OutputIncidentData]
+    removed: List[OutputIncidentData]
+    added_counter: Dict[str, int]
+    removed_counter: Dict[str, int]
+
+    def sort_key(self) -> List[int]:
+        return list(self.added_counter.values()) + list(self.removed_counter.values())
+
+
+def populate_incidents_list(
+    result_list: List[SCALocationVulnerability],
+    locations_with_incidents: Dict[str, OutputLocationData],
+    incident_list_key: str,
+    counter_key: str,
+) -> None:
+    for vuln_dep_file in result_list:
+        if vuln_dep_file.location not in locations_with_incidents:
+            locations_with_incidents[vuln_dep_file.location] = OutputLocationData(
+                location=vuln_dep_file.location,
+                added=[],
+                removed=[],
+                added_counter={"critical": 0, "high": 0, "medium": 0, "low": 0},
+                removed_counter={"critical": 0, "high": 0, "medium": 0, "low": 0},
+            )
+
+        # Update counter of vulns to sort locations by number of incidents
+        for pkg_version_with_vuln in vuln_dep_file.package_vulns:
+            for vuln in pkg_version_with_vuln.vulns:
+                incidents_list = locations_with_incidents[
+                    vuln_dep_file.location
+                ].__getattribute__(incident_list_key)
+                incidents_list.append(
+                    OutputIncidentData(
+                        package_full_name=pkg_version_with_vuln.package_full_name,
+                        version=pkg_version_with_vuln.version,
+                        severity=vuln.severity,
+                        summary=vuln.summary,
+                        fixed_version=vuln.fixed_version,
```

Comment:
Can any of these fields be None? for example this one or `cve_ids`. We might need some default values for `OutputIncidentData`?

Reply:
We don't as the default values are already populated in the `SCAVulnerability` model, so here the value always exist, although it can be `None`